cdap-docs/reference-manual/source/http-restful-api/http-restful-api-v3/preferences.rst

```diff
@@ -0,0 +1,163 @@
+
+.. list-table::
+:widths: 20 80
+   :header-rows: 1
+
+     * - Parameter
+       - Description
+     * - ``<namespace-id>``
+     - Namespace ID
+     * - ``<app-id>``
+     - Application ID
+     * - ``<program-type>``
+     - Program Type (flows, services, etc)
+     * - ``<program-id>``
+     - Program ID
+
+Preferences can be set only for entities that exist. For example, Preferences cannot be set for a Namespace
+that does not exist or an application that has not be deployed. Properties, which are map of string-string pairs are
+passed in the JSON request body.
+
+.. rubric:: HTTP Responses
+
+.. list-table::
+:widths: 20 80
+   :header-rows: 1
+
+     * - Status Codes
+       - Description
+     * - ``200 OK``
+     - The event successfully called the method, and the preferences was set
+     * - ``400 BAD REQUEST``
+     - The JSON body has invalid format
+     * - ``404 NOT FOUND``
+     - The entity, for which Preferences is being set, is not present
+
+
+Get Preferences
+---------------
+
+To get the preferences, issue an HTTP GET request::
+
+  GET http://<host>:<port>/v3/configuration/preferences/
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>/apps/<app-id>
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>/apps/<app-id>/<program-type>/<program-id>
+
+This will return a JSON String map that has the preferences in a JSON map format::
+
+  {"key1":"value1", "key2":"value2"}
+
+In order to get Resolved Preferences (collapsing Preferences from higher levels), one can set the ``resolved`` query parameter to ``true``::
+
+  GET http://<host>:<port>/v3/configuration/preferences?resolved=true
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>?resolved=true
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>/apps/<app-id>?resolved=true
+
+  GET http://<host>:<port>/v3/configuration/preferences/namespaces/<namespace-id>/apps/<app-id>/<program-type>/<program-id>?resolved=true
+
+.. list-table::
+:widths: 20 80
+   :header-rows: 1
+
+       * - Parameter
+         - Description
+       * - ``<namespace-id>``
+     - Namespace ID
+     * - ``<app-id>``
+     - Application ID
+     * - ``<program-type>``
+     - Program Type (flows, services, etc)
+     * - ``<program-id>``
+     - Program ID
+
+.. rubric:: HTTP Responses
+
+.. list-table::
+:widths: 20 80
+   :header-rows: 1
+
+       * - Status Codes
+         - Description
+       * - ``200 OK``
+     - The event successfully called the method, and the preferences was retrieved
+     * - ``404 NOT FOUND``
+     - The entity, for which Preferences is being retrieved, is not present
+
+Delete Preferences
+------------------
+To delete preferences, issue an HTTP DELETE::
```

Comment:
Can you delete the resolved preferences? Say, issue a delete at the namespace level that deletes the namespace and above preferences, but leaves everything below? Or do you have to delete them all manually (each level individually)?
We should mention what to do...